rstar/src/primitives/geomwithdata.rs

```diff
@@ -0,0 +1,163 @@
+/// into an r-tree. This struct removes some of the boilerplate required to do so.
+///
+/// **Note:** while the container itself implements [RTreeObject], you will have to go through its
+/// `geom` method in order to access geometry-specific methods.
+///
+/// # Examples
+/// ```
+/// use rstar::{RTree, PointDistance};
+/// use rstar::primitives::GeomWithData;
+///
+/// type RestaurantLocation = GeomWithData<&'static str, [f64; 2]>;
+///
+/// let mut restaurants = RTree::new();
+/// restaurants.insert(RestaurantLocation::new("Pete's Pizza Place", [0.3, 0.2]));
+/// restaurants.insert(RestaurantLocation::new("The Great Steak", [-0.8, 0.0]));
+/// restaurants.insert(RestaurantLocation::new("Fishy Fortune", [0.2, -0.2]));
+///
+/// let my_location = [0.0, 0.0];
+///
+/// // Now find the closest restaurant!
+/// let place = restaurants.nearest_neighbor(&my_location).unwrap();
+/// println!("Let's go to {}", place.data);
+/// println!("It's really close, only {} miles", place.distance_2(&my_location));
+/// ```
+/// ```
+/// // We can use any primitive type
+/// use rstar::{RTree, PointDistance};
+/// use rstar::primitives::GeomWithData;
+/// use rstar::primitives::Line;
+///
+/// let mut lines = RTree::new();
+/// lines.insert(GeomWithData::new(
+///     "Pete's Pizza Place",
+///     Line::new([0.0, 0.0], [1.0, 1.0]),
+/// ));
+/// lines.insert(GeomWithData::new(
+///     "The Great Steak",
+///     Line::new([0.0, 0.0], [-1.0, 1.0]),
+/// ));
+///
+/// let my_location = [0.0, 0.0];
+///
+/// // Now find the closest line
+/// let place = lines.nearest_neighbor(&my_location).unwrap();
+/// println!("Let's go to {}", place.data);
+/// println!(
+///     "It's really close, only {} miles",
+///     place.geom().distance_2(&my_location)
+/// );
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd)]
+pub struct GeomWithData<T, R> {
+    geom: R,
+    /// Data to be associated with the geometry being stored in the RTree
+    pub data: T,
+}
+
+impl<T, R: RTreeObject> RTreeObject for GeomWithData<T, R> {
+    type Envelope = R::Envelope;
+
+    fn envelope(&self) -> Self::Envelope {
+        self.geom.envelope()
+    }
+}
+
+impl<T, R: PointDistance> PointDistance for GeomWithData<T, R> {
+    fn distance_2(
+        &self,
+        point: &<Self::Envelope as Envelope>::Point,
+    ) -> <<Self::Envelope as Envelope>::Point as Point>::Scalar {
+        self.geom.distance_2(point)
+    }
+
+    fn contains_point(&self, p: &<Self::Envelope as Envelope>::Point) -> bool {
+        self.geom.contains_point(p)
+    }
+
+    fn distance_2_if_less_or_equal(
+        &self,
+        point: &<Self::Envelope as Envelope>::Point,
+        max_distance_2: <<Self::Envelope as Envelope>::Point as Point>::Scalar,
+    ) -> Option<<<Self::Envelope as Envelope>::Point as Point>::Scalar> {
+        self.geom.distance_2_if_less_or_equal(point, max_distance_2)
+    }
+}
+
+impl<T, P> GeomWithData<T, P> {
```

Comment:
Nittest of nits, but it might be confusing that the type parameter is called `P` here but `R` everywhere else. Also, you can construct the instance using `Self` as well, i.e.
```
pub fn new(data: T, geom: R) -> Self {
  Self { geom, data }
}
```

Reply:
Oops, totally legit nit. Fixed!